File: common/mrtjp/projectred/transportation/TransportationRecipes.java
package mrtjp.projectred.transportation;

import cpw.mods.fml.common.registry.GameRegistry;
import mrtjp.projectred.ProjectRedTransportation;
import mrtjp.projectred.core.ItemPart.EnumPart;
import mrtjp.projectred.core.PRColors;
import mrtjp.projectred.transportation.EnumRoutingChip.ChipVal;
import net.minecraft.block.Block;
import net.minecraft.inventory.InventoryCrafting;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.IRecipe;
import net.minecraft.world.World;
import net.minecraftforge.oredict.ShapedOreRecipe;


public class TransportationRecipes
{
    public static void initRecipes()
    {
        initPipeRecipes();
        initChipRecipes();
        initUpgradeRecipes();
    }

    private static void initPipeRecipes()
    {
        /** Item Transport Pipe **/
        GameRegistry.addRecipe(PipeDef.BASIC().getItemStack(16),
                "sgs",
                'g', Block.thinGlass,
                's', Block.stone
        );

        /** Routed Junction Pipe **/
        GameRegistry.addRecipe(PipeDef.ROUTEDJUNCTION().getItemStack(16),
                "RrG",
                "dgd",
                "GrR",
                'R', EnumPart.REDILLUMAR.getItemStack(),
                'r', EnumPart.INFUSEDSILICON.getItemStack(),
                'G', EnumPart.GREENILLUMAR.getItemStack(),
                'd', Item.diamond,
                'g', Block.thinGlass
        );

        /** Routed Interface Pipe **/
        GameRegistry.addRecipe(PipeDef.ROUTEDINTERFACE().getItemStack(),
                "rgr",
                "gjg",
                "rgr",
                'g', Item.goldNugget,
                'j', PipeDef.ROUTEDJUNCTION().getItemStack(),
                'r', Item.redstone
        );

        /** Routed Crafting Pipe **/
        GameRegistry.addRecipe(PipeDef.ROUTEDCRAFTING().getItemStack(),
                "rgr",
                "rjr",
                "rgr",
                'r', Item.redstone,
                'g', Item.glowstone,
                'j', PipeDef.ROUTEDJUNCTION().getItemStack()
        );

        /** Routed Request Pipe **/
        GameRegistry.addRecipe(PipeDef.ROUTEDREQUEST().getItemStack(),
                "rdr",
                "rjr",
                "rdr",
                'r', Item.redstone,
                'd', Item.diamond,
                'j', PipeDef.ROUTEDJUNCTION().getItemStack()
        );

        /** Routed Extension Pipe **/
        GameRegistry.addRecipe(PipeDef.ROUTEDEXTENSION().getItemStack(),
                " r ",
                "rjr",
                " r ",
                'r', Item.redstone,
                'j', PipeDef.ROUTEDJUNCTION().getItemStack()
        );

        /** Routed Firewall Pipe **/
        GameRegistry.addRecipe(PipeDef.ROUTEDFIREWALL().getItemStack(),
                "bcb",
                "cjc",
                "bcb",
                'b', Block.netherBrick,
                'c', Item.magmaCream,
                'j', PipeDef.ROUTEDJUNCTION().getItemStack()
        );
    }

    private static void initChipRecipes()
    {
        /** Chip reset **/
        for (ChipVal r : EnumRoutingChip.VALID_CHIPS())
            GameRegistry.addRecipe(new IRecipe()
            {

                @Override
                public boolean matches(InventoryCrafting inv, World world)
                {
                    return getCraftingResult(inv) != null;
                }

                @Override
                public ItemStack getCraftingResult(InventoryCrafting inv)
                {
                    ChipVal type = getType(inv);
                    if (type != null)
                        if (isTypeExclusive(type, inv))
                            return type.getItemStack(countUnits(inv));

                    return null;
                }

                public ChipVal getType(InventoryCrafting inv)
                {
                    for (int i = 0; i < inv.getSizeInventory(); i++)
                    {
                        ChipVal type = EnumRoutingChip.getForStack(inv.getStackInSlot(i));
                        if (type != null)
                            return type;
                    }
                    return null;
                }

                public boolean isTypeExclusive(ChipVal type, InventoryCrafting inv)
                {
                    for (int i = 0; i < inv.getSizeInventory(); i++)
                    {
                        ItemStack stack = inv.getStackInSlot(i);
                        if (stack != null && !(stack.getItem() instanceof ItemRoutingChip))
                            return false;

                        ChipVal type2 = EnumRoutingChip.getForStack(stack);
                        if (type2 != null && !type2.equals(type))
                            return false;
                    }
                    return true;
                }

                public int countUnits(InventoryCrafting inv)
                {
                    int count = 0;
                    for (int i = 0; i < inv.getSizeInventory(); i++)
                        if (inv.getStackInSlot(i) != null)
                            count++;

                    return count;
                }

                @Override
                public int getRecipeSize()
                {
                    return 2;
                }

                @Override
                public ItemStack getRecipeOutput()
                {
                    return null;
                }

            });

        /** Null chip **/
        GameRegistry.addRecipe(EnumPart.NULLROUTINGCHIP.getItemStack(),
                "gpp",
                "grr",
                "g  ",
                'g', Item.goldNugget,
                'p', Item.paper,
                'r', Item.redstone
        );

        /** Item Responder **/
        addChipRecipe(EnumRoutingChip.ITEMRESPONDER().getItemStack(),
                Item.ingotIron, Item.redstone, Item.redstone,
                EnumPart.ORANGEILLUMAR.getItemStack(),
                EnumPart.ORANGEILLUMAR.getItemStack()
        );

        /** Dynamic Item Responder **/
        addChipRecipe(EnumRoutingChip.DYNAMICITEMRESPONDER().getItemStack(),
                Item.ingotIron, Item.redstone, EnumPart.CYANILLUMAR.getItemStack(),
                EnumPart.ORANGEILLUMAR.getItemStack(),
                EnumPart.ORANGEILLUMAR.getItemStack()
        );

        /** Item Overflow Responder **/
        addChipRecipe(EnumRoutingChip.ITEMOVERFLOWRESPONDER().getItemStack(),
                Item.ingotIron, Item.redstone, Item.redstone,
                EnumPart.GREENILLUMAR.getItemStack(),
                EnumPart.GREENILLUMAR.getItemStack()
        );

        /** Item Terminator **/
        addChipRecipe(EnumRoutingChip.ITEMTERMINATOR().getItemStack(),
                Item.ingotIron, Item.redstone, Item.redstone,
                EnumPart.PURPLEILLUMAR.getItemStack(),
                EnumPart.GREYILLUMAR.getItemStack()
        );

        /** Item Extractor **/
        addChipRecipe(EnumRoutingChip.ITEMEXTRACTOR().getItemStack(),
                Item.ingotIron, Item.redstone, Item.redstone,
                EnumPart.CYANILLUMAR.getItemStack(),
                EnumPart.CYANILLUMAR.getItemStack()
        );

        /** Item Broadcaster **/
        addChipRecipe(EnumRoutingChip.ITEMBROADCASTER().getItemStack(),
                Item.ingotGold, Item.redstone, Item.redstone,
                EnumPart.MAGENTAILLUMAR.getItemStack(),
                EnumPart.MAGENTAILLUMAR.getItemStack()
        );

        /** Item Stock Keeper **/
        addChipRecipe(EnumRoutingChip.ITEMSTOCKKEEPER().getItemStack(),
                Item.diamond, Item.redstone, Item.redstone,
                EnumPart.BLUEILLUMAR.getItemStack(),
                EnumPart.BLUEILLUMAR.getItemStack()
        );

        /** Item Crafting **/
        addChipRecipe(EnumRoutingChip.ITEMCRAFTING().getItemStack(),
                Item.glowstone, Item.redstone, Item.glowstone,
                EnumPart.LIMEILLUMAR.getItemStack(),
                EnumPart.LIMEILLUMAR.getItemStack()
        );
    }

    private static void addChipRecipe(ItemStack result, Object bus, Object material2, Object material1, Object dyeLeft, Object dyeRight)
    {
        GameRegistry.addRecipe(new ShapedOreRecipe(result,
                "dMD",
                "bcm",
                "dMD",
                'd', dyeLeft,
                'M', material2,
                'D', dyeRight,
                'b', bus,
                'c', EnumPart.NULLROUTINGCHIP.getItemStack(),
                'm', material1
        ));
    }

    private static void initUpgradeRecipes()
    {
        /** Router Utility **/
        GameRegistry.addRecipe(new ItemStack(ProjectRedTransportation.itemRouterUtility()),
                "  r",
                "iei",
                "iii",
                'r', Item.redstone,
                'i', Item.ingotIron,
                'e', Item.emerald
        );

        /** Null Upgrade **/
        GameRegistry.addRecipe(EnumPart.NULLUPGRADECHIP.getItemStack(),
                "prp",
                "rrr",
                "prp",
                'p', Item.paper,
                'r', Item.redstone
        );

        /** LX **/
        GameRegistry.addRecipe(EnumPart.CHIPUPGRADE_LX.getItemStack(),
                "rrr",
                " ng",
                "r r",
                'r', Item.redstone,
                'n', EnumPart.NULLUPGRADECHIP.getItemStack(),
                'g', Item.goldNugget
        );

        /** RX **/
        GameRegistry.addRecipe(EnumPart.CHIPUPGRADE_RX.getItemStack(),
                "r r",
                "gn ",
                "rrr",
                'r', Item.redstone,
                'n', EnumPart.NULLUPGRADECHIP.getItemStack(),
                'g', Item.goldNugget
        );

        /** LY **/
        GameRegistry.addRecipe(EnumPart.CHIPUPGRADE_LY.getItemStack(),
                "l l",
                " n ",
                " l ",
                'l', PRColors.BLUE.getDye(),
                'n', EnumPart.CHIPUPGRADE_LX.getItemStack()
        );

        /** RY **/
        GameRegistry.addRecipe(EnumPart.CHIPUPGRADE_RY.getItemStack(),
                "l l",
                " n ",
                " l ",
                'l', PRColors.BLUE.getDye(),
                'n', EnumPart.CHIPUPGRADE_RX.getItemStack()
        );

        /** LZ **/
        GameRegistry.addRecipe(EnumPart.CHIPUPGRADE_LZ.getItemStack(),
                "r r",
                " n ",
                "rer",
                'r', Item.redstone,
                'n', EnumPart.CHIPUPGRADE_LY.getItemStack(),
                'e', Item.emerald
        );

        /** RZ **/
        GameRegistry.addRecipe(EnumPart.CHIPUPGRADE_RZ.getItemStack(),
                "r r",
                " n ",
                "rer",
                'r', Item.redstone,
                'n', EnumPart.CHIPUPGRADE_RY.getItemStack(),
                'e', Item.emerald
        );
    }
}

File: src/mrtjp/projectred/transportation/TransportationRecipes.scala
package mrtjp.projectred.transportation

import java.lang.{Character => JC}

import mrtjp.projectred.ProjectRedTransportation
import mrtjp.projectred.core.PartDefs
import mrtjp.projectred.transportation.RoutingChipDefs.ChipVal
import net.minecraft.init.{Blocks, Items}
import net.minecraft.inventory.InventoryCrafting
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.IRecipe
import net.minecraft.world.World
import net.minecraftforge.common.ForgeHooks
import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraftforge.oredict.RecipeSorter.Category._
import net.minecraftforge.oredict.{RecipeSorter, ShapedOreRecipe}

object TransportationRecipes
{
    def initRecipes()
    {
        initNetPipeRecipes()
        initPressureTubeRecipes()
        initChipRecipes()
        initUpgradeRecipes()
        initMiscRecipes()
    }

    private def initNetPipeRecipes()
    {
        /** Item Transport pipe **/
        GameRegistry.addRecipe(new ShapedOreRecipe(PipeDefs.BASIC.makeStack(16),
            "sgs",
            'g':JC, "paneGlassColorless",
            's':JC, "stone"
        ))

        /** Routed Junction pipe **/
        GameRegistry.addRecipe(new ShapedOreRecipe(PipeDefs.ROUTEDJUNCTION.makeStack(16),
            "RrG",
            "dgd",
            "GrR",
            'R':JC, PartDefs.REDILLUMAR.makeStack,
            'r':JC, PartDefs.INFUSEDSILICON.makeStack,
            'G':JC, PartDefs.GREENILLUMAR.makeStack,
            'd':JC, "gemDiamond",
            'g':JC, "paneGlassColorless"
        ))

        /** Routed Interface Pipe **/
        GameRegistry.addRecipe(new ShapedOreRecipe(PipeDefs.ROUTEDINTERFACE.makeStack,
            "rgr",
            "gjg",
            "rgr",
            'g':JC, "nuggetGold",
            'j':JC, PipeDefs.ROUTEDJUNCTION.makeStack,
            'r':JC, "dustRedstone"
        ))

        /** Routed Request Pipe **/
        GameRegistry.addRecipe(new ShapedOreRecipe(PipeDefs.ROUTEDREQUEST.makeStack,
            "rdr", "rjr", "rdr",
            'r':JC, "dustRedstone",
            'd':JC, "gemDiamond",
            'j':JC, PipeDefs.ROUTEDJUNCTION.makeStack
        ))

        /** Routed Firewall Pipe **/
        GameRegistry.addRecipe(PipeDefs.ROUTEDFIREWALL.makeStack,
            "bcb", "cjc", "bcb",
            'b':JC, Blocks.NETHER_BRICK,
            'c':JC, Items.MAGMA_CREAM,
            'j':JC, PipeDefs.ROUTEDJUNCTION.makeStack
        )

        /** Network Valve pipe **/
        GameRegistry.addRecipe(PipeDefs.NETWORKVALVE.makeStack,
            " g ", "gbg", " l ",
            'g':JC, Blocks.GLASS_PANE,
            'b':JC, PipeDefs.BASIC.makeStack,
            'l':JC, Blocks.LEVER
        )

        /** Network Latency pipe **/
        GameRegistry.addRecipe(PipeDefs.NETWORKLATENCY.makeStack,
            " c ", "cbc", " c ",
            'c':JC, Blocks.COBBLESTONE,
            'b':JC, PipeDefs.BASIC.makeStack
        )
    }

    private def initPressureTubeRecipes()
    {
        /** Pressure Tube **/
        GameRegistry.addRecipe(new ShapedOreRecipe(PipeDefs.PRESSURETUBE.makeStack(8),
            "gpg", "p p", "gpg",
            'g':JC, "nuggetGold",
            'p':JC, Blocks.GLASS_PANE
        ))

        /** Pressure Resistance Tube **/
        GameRegistry.addRecipe(new ShapedOreRecipe(PipeDefs.RESISTANCETUBE.makeStack,
            "i", "t",
            'i':JC, "ingotIron",
            't':JC, PipeDefs.PRESSURETUBE.makeStack
        ))
    }

    private def initChipRecipes()
    {
        RecipeSorter.register("projectred:chipreset", classOf[ChipResetRecipe], SHAPED, "after:forge:shaped")

        /** Chip reset **/
        GameRegistry.addRecipe(new ChipResetRecipe)

        /** Null chip **/
        GameRegistry.addRecipe(new ShapedOreRecipe(PartDefs.NULLROUTINGCHIP.makeStack,
            "gpp", "grr", "g  ",
            'g':JC, "nuggetGold",
            'p':JC, Items.PAPER,
            'r':JC, "dustRedstone"
        ))

        /** Item Responder **/
        addChipRecipe(RoutingChipDefs.ITEMRESPONDER.makeStack,
            "ingotIron", "dustRedstone", "dustRedstone",
            PartDefs.ORANGEILLUMAR.makeStack,
            PartDefs.ORANGEILLUMAR.makeStack)

        /** Dynamic Item Responder **/
        addChipRecipe(RoutingChipDefs.DYNAMICITEMRESPONDER.makeStack,
            "ingotIron", "dustRedstone",
            PartDefs.CYANILLUMAR.makeStack,
            PartDefs.ORANGEILLUMAR.makeStack,
            PartDefs.ORANGEILLUMAR.makeStack)

        /** Item Overflow Responder **/
        addChipRecipe(RoutingChipDefs.ITEMOVERFLOWRESPONDER.makeStack,
            "ingotIron", "dustRedstone", "dustRedstone",
            PartDefs.GREENILLUMAR.makeStack,
            PartDefs.GREENILLUMAR.makeStack)

        /** Item Terminator **/
        addChipRecipe(RoutingChipDefs.ITEMTERMINATOR.makeStack,
            "ingotIron", "dustRedstone", "dustRedstone",
            PartDefs.PURPLEILLUMAR.makeStack,
            PartDefs.GREYILLUMAR.makeStack)

        /** Item Extractor **/
        addChipRecipe(RoutingChipDefs.ITEMEXTRACTOR.makeStack,
            "ingotIron", "dustRedstone", "dustRedstone",
            PartDefs.CYANILLUMAR.makeStack,
            PartDefs.CYANILLUMAR.makeStack)

        /** Item Broadcaster **/
        addChipRecipe(RoutingChipDefs.ITEMBROADCASTER.makeStack,
            "ingotGold", "dustRedstone", "dustRedstone",
            PartDefs.MAGENTAILLUMAR.makeStack,
            PartDefs.MAGENTAILLUMAR.makeStack)

        /** Item Stock Keeper **/
        addChipRecipe(RoutingChipDefs.ITEMSTOCKKEEPER.makeStack,
            "gemDiamond", "dustRedstone", "dustRedstone",
            PartDefs.BLUEILLUMAR.makeStack,
            PartDefs.BLUEILLUMAR.makeStack)

        /** Item Crafting **/
        addChipRecipe(RoutingChipDefs.ITEMCRAFTING.makeStack,
            "dustGlowstone", "dustRedstone", "dustGlowstone",
            PartDefs.LIMEILLUMAR.makeStack,
            PartDefs.LIMEILLUMAR.makeStack)

        /** Item Crafting Extension **/
        addChipRecipe(RoutingChipDefs.ITEMEXTENSION.makeStack,
            "dustGlowstone", "dustRedstone", "dustRedstone",
            PartDefs.REDILLUMAR.makeStack,
            PartDefs.REDILLUMAR.makeStack)


        def addChipRecipe(result:ItemStack, bus:AnyRef, material2:AnyRef, material1:AnyRef, dyeLeft:AnyRef, dyeRight:AnyRef)
        {
            GameRegistry.addRecipe(new ShapedOreRecipe(result,
                "dMD", "bcm", "dMD",
                'd':JC, dyeLeft,
                'M':JC, material2,
                'D':JC, dyeRight,
                'b':JC, bus,
                'c':JC, PartDefs.NULLROUTINGCHIP.makeStack,
                'm':JC, material1))
        }
    }

    private def initUpgradeRecipes()
    {
        /** Router Utility **/
        GameRegistry.addRecipe(new ShapedOreRecipe(new ItemStack(ProjectRedTransportation.itemRouterUtility),
            "  r", "iei", "iii",
            'r':JC, "dustRedstone",
            'i':JC, "ingotIron",
            'e':JC, "dyeGreen"))
    }

    def initMiscRecipes()
    {
    }
}

class ChipResetRecipe extends IRecipe
{
    override def getRemainingItems(inv:InventoryCrafting) = ForgeHooks.defaultRecipeGetRemainingItems(inv)

    def matches(inv:InventoryCrafting, world:World) = !getCraftingResult(inv).isEmpty

    def getCraftingResult(inv:InventoryCrafting):ItemStack =
    {
        val cdef = getType(inv)
        if (cdef != null) if (isTypeExclusive(cdef, inv)) return cdef.makeStack(countUnits(inv))
        null
    }

    def getType(inv:InventoryCrafting):ChipVal =
    {
        for (i <- 0 until inv.getSizeInventory)
        {
            val cdef = RoutingChipDefs.getForStack(inv.getStackInSlot(i))
            if (cdef != null) return cdef
        }
        null
    }

    def isTypeExclusive(cdef:ChipVal, inv:InventoryCrafting):Boolean =
    {
        for (i <- 0 until inv.getSizeInventory)
        {
            val stack = inv.getStackInSlot(i)
            if (!stack.isEmpty && !stack.getItem.isInstanceOf[ItemRoutingChip]) return false
            val type2 = RoutingChipDefs.getForStack(stack)
            if (type2 != null && !(type2 == cdef)) return false
        }
        true
    }

    def countUnits(inv:InventoryCrafting):Int =
    {
        var count = 0
        for (i <- 0 until inv.getSizeInventory)
            if (!inv.getStackInSlot(i).isEmpty) count += 1
        count
    }

    def getRecipeSize = 2
    def getRecipeOutput = ItemStack.EMPTY
}
